### ISSUE TITLE
Feature Request: Add assembly for references from Assembly objects

### ISSUE THREAD

**Issue by scegg** [dotnet/roslyn#21579]
In the current version of Roslyn, the assemblies for references can only be provided from file name or PE-image, but not from an assembly object.

Is that possible to add support to pass the reference by an existed assembly object?

Note: Some assembly, which is generated in memory, may not have a file name.

**Comment by jcouv** [comment]
@scegg The compiler APIs already allow doing this. For example, consider this [test](https://github.com/dotnet/roslyn/blob/master/src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTupleTest.cs#L282) (`ConstraintAttributesAreNotSharedAcrossTypeRefs`) which compiles a snippet of C# in memory, emits the assembly to a bit array in memory, then uses that as a reference for a second compilation.
Look at the code for the internal helper method `EmitToImageReference` for more details.

**Comment by scegg** [comment]
@jcouv Thanks for your reply.
But I guess I doesn't get your point.

A case for demo:
One user need to compile a part of C# code using Roslyn.
class Demo
{
  public Extreme MyProperty {get;set;}
}

Extreme is defined in another assembly, which is generated by another project in memory (only have an assembly object, but no PE-image, nor filename), and loaded into this app domain already.
The user can use typeof(Extreme).Assembly to get the assembly but no way to get the PE-image (bit array), nor filename.

How to call Roslyn to build this?

**Comment by jcouv** [comment]
@scegg Thanks for the clarification. I had misunderstood your scenario. You're trying to use types that only exist inside the runtime, and you have an `Assembly` reflection object. 
Indeed, I don't think it is possible to use such type from a Roslyn `Compilation`.

I'm trying to think how this could be done, but I'm falling short. In particular, the compiler emits metadata which has assembly references. How could the emitted metadata reference an assembly that doesn't exist and has no identity?

Tagging @jaredpar in case he has an idea how that might work.

**Comment by tmat** [comment]
@scegg On Core CLR you can use [AssemblyExtensions.TryGetRawMetadata](https://docs.microsoft.com/en-us/dotnet/api/system.reflection.metadata.assemblyextensions.trygetrawmetadata?view=netcore-2.0#System_Reflection_Metadata_AssemblyExtensions_TryGetRawMetadata_System_Reflection_Assembly_System_Byte___System_Int32__)

**Comment by scegg** [comment]
@tmat Thanks. Actually, the main problem is related to dotnet framework instead of core. In core, i have trusted platform at least.

**Comment by tmat** [comment]
@scegg Can you change the project that generates the assembly in a way that allows you to retrieve the byte[] for the emitted assembly?

**Comment by scegg** [comment]
@tmat Those assemblies are not generated by me, so...

**Comment by tmat** [comment]
@scegg I agree it's not ideal. Unfortunately, I don't have a better solution for you at this point. It is possible that in future ```AssemblyExtensions.TryGetRawMetadata``` will be ported to .NET Framework. But there is no ETA for that currently. Updating Roslyn to be able to emit against System.Reflection objects is a lot of work and unlikely to happen in foreseeable future.
